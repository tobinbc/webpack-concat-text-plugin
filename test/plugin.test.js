/* eslint-disable max-len */
/* eslint-disable no-shadow */
/* eslint-disable max-nested-callbacks */
import fs from "fs";
import path from "path";

jest.unmock("concat");
jest.unmock("glob");

/* eslint-disable import/imports-first */
import { compile, createCompiler, cleanErrorStack } from "./utils";
import ConcatTextPlugin, {
	getExtFromGlobPath,
	getRelativeTargetPath,
	globTextFiles,
	mergeWithOutputOptions,
	PLUGIN_NAME } from "../src/index";
/* eslint-enable */

const cases = process.env.CASES ? process.env.CASES.split(",") : fs.readdirSync(path.join(__dirname, "cases"));

describe(PLUGIN_NAME, () => {
	cases.forEach((testCase) => {
		it(testCase, (done) => {

			const configFile = path.resolve("test", "cases", testCase, "config.js");
			const filename = `${testCase}.js`;

			const caseConfig = (fs.existsSync(configFile))
				? require(configFile) // eslint-disable-line global-require
				: {};

			const pluginConfig = Object.assign({}, caseConfig);
			const webpackConfig = {
				output: {
					filename,
					path: path.join(__dirname, "__output__", testCase, "dist")
				}
			};

			const { compiler, memfs } = createCompiler(webpackConfig);

			new ConcatTextPlugin(pluginConfig).apply(compiler);

			compile(compiler).then((stats) => {
				const errors = stats.compilation.errors.map(cleanErrorStack);
				const warnings = stats.compilation.warnings.map(cleanErrorStack);
				const files = stats.toJson().assets.filter((asset) => asset.name !== filename);

				expect(errors).toMatchSnapshot("errors");
				expect(warnings).toMatchSnapshot("warnings");

				files.forEach((asset) => {
					if (Object.prototype.hasOwnProperty.call(stats.compilation.assets, asset.name)) {
						// Covers previous path test
						const sourcePath = path.join(__dirname, "__output__", testCase, "dist", asset.name);
						const source = memfs.readFileSync(sourcePath, "utf8");
						expect(source).toMatchSnapshot("source");

					}
				});

				done();
			});
		});
	});

	describe("Utility Functions", () => {

		it("can extract file type extension from glob path", () => {
			[ // Test Cases
				{ globPath: "test/*.js", expected: ".js" },
				{ globPath: "test/**/*.ts", expected: ".ts" },
				{ globPath: "test/(res|assets)/*.md", expected: ".md" },
				{ globPath: "test/{res|assets}/*.xml", expected: ".xml" }
			].forEach(({ globPath, expected }) => {
				const actual = getExtFromGlobPath(globPath);
				expect(actual).toBe(expected);
			});
		});

		it("can extract empty file type extension", () => {
			[ // Test Cases
				"test/*.test.{js,ts}",
				"test/*.js?(x)",
				"test/*.*sx",
				"test/*.[A-Z]",
				"test/*.*(htm|html)",
				"test/*.+(md|markdown)"
			].map(getExtFromGlobPath).forEach((expected) => {
				expect(expected).toHaveLength(0);
			});
		});

		it("can get relative target path", () => {
			const startingPath = path.join(__dirname, "test");

			[ // Test Cases
				{ outputPath: "cases", expected: "cases" },
				{ outputPath: "../test", expected: "../test" },
				{ outputPath: "..", expected: ".." },
				{ outputPath: __dirname, expected: ".." },
				{ outputPath: path.join(__dirname, "test", "coverage"), expected: "coverage" },
				{ outputPath: path.join(__dirname, "test", "cases", "foo"), expected: path.join("cases", "foo") },
				{ outputPath: path.join(__dirname, "test"), expected: "" },
				{ outputPath: path.join(__dirname, "dist", "test"), expected: path.join("..", "dist", "test") }
			].forEach(({ outputPath, expected }) => {
				const actual = getRelativeTargetPath(startingPath, outputPath);
				expect(actual).toBe(expected);
			});
		});

		it("can merge plugin options with compiler output options", () => {
			const outputOptions = { filename: "main.js", path: path.join(__dirname, "dist") };

			[ // Test Cases
				{
					pluginOptions: { files: "src/**/*.properties", name: "values.properties" },
					expected: { name: "values.properties", outputPath: path.join(__dirname, "dist") }
				},
				{
					pluginOptions: { files: "src/**/*.properties", outputPath: "../" },
					expected: { name: "main.properties", outputPath: "../" }
				},
				{
					pluginOptions: { files: "src/**/*.properties", name: "res.props", outputPath: "res" },
					expected: { name: "res.props", outputPath: "res" }
				},
				{
					pluginOptions: { files: "src/**/*.properties" },
					expected: { name: "main.properties", outputPath: path.join(__dirname, "dist") }
				}
			].forEach(({ pluginOptions, expected }) => {
				const actual = mergeWithOutputOptions(outputOptions, pluginOptions);
				expect(actual).toEqual(Object.assign({}, expected, { files: pluginOptions.files }));
			});
		});

		it("can glob files asynchronously", () => {
			return globTextFiles(`test/cases/{${cases.join(",")},}/*.js`)
				.then((actual) => {
					expect(actual).toHaveLength(cases.length);
				});
		});

	});

});
