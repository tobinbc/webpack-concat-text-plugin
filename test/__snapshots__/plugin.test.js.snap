// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConcatTextPlugin default-name: errors 1`] = `Array []`;

exports[`ConcatTextPlugin default-name: source 1`] = `
"TestProperty.Value=Hello World

TestProperty2.Value=Hallo Welt
"
`;

exports[`ConcatTextPlugin default-name: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin default-name-magic-glob-ext: errors 1`] = `Array []`;

exports[`ConcatTextPlugin default-name-magic-glob-ext: source 1`] = `
"<h1>Hello World</h1>

<h2>Hello World</h2>
"
`;

exports[`ConcatTextPlugin default-name-magic-glob-ext: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin default-name-no-glob-ext: errors 1`] = `Array []`;

exports[`ConcatTextPlugin default-name-no-glob-ext: source 1`] = `
"module.exports = {
	files: \`\${__dirname}/*\`,
	outputPath: __dirname.replace(\\"cases\\", \\"__output__\\")
};

TestProperty.Value=Hello World

TestProperty2.Value=Hallo Welt
"
`;

exports[`ConcatTextPlugin default-name-no-glob-ext: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin default-output-path: errors 1`] = `Array []`;

exports[`ConcatTextPlugin default-output-path: source 1`] = `
"TestProperty.Value=Hello World

TestProperty2.Value=Hallo Welt
"
`;

exports[`ConcatTextPlugin default-output-path: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin empty: errors 1`] = `Array []`;

exports[`ConcatTextPlugin empty: source 1`] = `
"
"
`;

exports[`ConcatTextPlugin empty: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin multiple-file-types: errors 1`] = `Array []`;

exports[`ConcatTextPlugin multiple-file-types: source 1`] = `
"TestProperty.Value=Hello World

Hallo Welt

# Hello World
"
`;

exports[`ConcatTextPlugin multiple-file-types: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin multiple-file-types-with-name: errors 1`] = `Array []`;

exports[`ConcatTextPlugin multiple-file-types-with-name: source 1`] = `
"TestProperty.Value=Hello World

Hallo Welt

# Hello World
"
`;

exports[`ConcatTextPlugin multiple-file-types-with-name: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin multiple-files: errors 1`] = `Array []`;

exports[`ConcatTextPlugin multiple-files: source 1`] = `
"TestProperty.Value=Hello World

TestProperty2.Value=Hallo Welt

TestProperty3.One.Value=Hallo Welt
TestProperty3.Two.Value=Foo
"
`;

exports[`ConcatTextPlugin multiple-files: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin nested-files: errors 1`] = `Array []`;

exports[`ConcatTextPlugin nested-files: source 1`] = `
"TestProperty.Value=Hello World

TestProperty3.One.Value=Hallo Welt
TestProperty3.Two.Value=Foo

TestProperty2.Value=Hallo Welt
"
`;

exports[`ConcatTextPlugin nested-files: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin no-trailing-line-end: errors 1`] = `Array []`;

exports[`ConcatTextPlugin no-trailing-line-end: source 1`] = `
"TestProperty.Value=Hello World
TestProperty2.Value=Hallo Welt"
`;

exports[`ConcatTextPlugin no-trailing-line-end: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin relative-files-path: errors 1`] = `Array []`;

exports[`ConcatTextPlugin relative-files-path: source 1`] = `
"TestProperty.Value=Hello World

TestProperty2.Value=Hallo Welt
"
`;

exports[`ConcatTextPlugin relative-files-path: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin relative-output-path: errors 1`] = `Array []`;

exports[`ConcatTextPlugin relative-output-path: source 1`] = `
"TestProperty.Value=Hello World

TestProperty2.Value=Hallo Welt
"
`;

exports[`ConcatTextPlugin relative-output-path: warnings 1`] = `Array []`;

exports[`ConcatTextPlugin webpack-entry-name: errors 1`] = `Array []`;

exports[`ConcatTextPlugin webpack-entry-name: source 1`] = `
"TestProperty.Value=Hello World

TestProperty2.Value=Hallo Welt
"
`;

exports[`ConcatTextPlugin webpack-entry-name: warnings 1`] = `Array []`;
